motionplan/tpSpaceRRT.go

```diff
@@ -172,185 +205,329 @@ func (mp *tpSpaceRRTMotionPlanner) planRunner(
 				rrt.solutionChan <- &rrtPlanReturn{steps: path, maps: rrt.maps}
 				return
 			}
 		}
+		if iter%mp.algOpts.attemptSolveEvery == 0 {
+			// Attempt a solve; we exhaustively iterate through our goal tree and attempt to find any connection to the seed tree
+			paths := [][]node{}
+			for goalMapNode := range rrt.maps.goalMap {
+				seedReached := mp.attemptExtension(ctx, goalMapNode, rrt.maps.startMap, false)
+				if seedReached.error != nil {
+					rrt.solutionChan <- &rrtPlanReturn{planerr: seedReached.error, maps: rrt.maps}
+					return
+				}
+				if seedReached.node == nil {
+					continue
+				}
+				reachedDelta := mp.planOpts.DistanceFunc(&ik.Segment{StartPosition: seedReached.node.Pose(), EndPosition: goalMapNode.Pose()})
+				if reachedDelta <= mp.algOpts.poseSolveDist {
+					// If we've reached the goal, extract the path from the RRT trees and return
+					path := extractPath(rrt.maps.startMap, rrt.maps.goalMap, &nodePair{a: seedReached.node, b: goalMapNode}, false)
+					paths = append(paths, path)
+				}
+			}
+			if len(paths) > 0 {
+				var bestPath []node
+				bestCost := math.Inf(1)
+				for _, goodPath := range paths {
+					currCost := sumCosts(goodPath)
+					if currCost < bestCost {
+						bestCost = currCost
+						bestPath = goodPath
+					}
+				}
+				correctedPath, err := rectifyTPspacePath(bestPath, mp.frame)
+				if err != nil {
+					rrt.solutionChan <- &rrtPlanReturn{planerr: err, maps: rrt.maps}
+					return
+				}
+				rrt.solutionChan <- &rrtPlanReturn{steps: correctedPath, maps: rrt.maps}
+				return
+			}
+		}
+
+		// Get random cartesian configuration
+		randPosNode, err = mp.sample(midptNode, iter+1)
+		if err != nil {
+			rrt.solutionChan <- &rrtPlanReturn{planerr: err, maps: rrt.maps}
+			return
+		}
 	}
 	rrt.solutionChan <- &rrtPlanReturn{maps: rrt.maps, planerr: errors.New("tpspace RRT unable to create valid path")}
 }
 
-// getExtensionCandidate will return either nil, or the best node on a PTG to reach the desired random node and its RRT tree parent.
+// getExtensionCandidate will return either nil, or the best node on a valid PTG to reach the desired random node and its RRT tree parent.
 func (mp *tpSpaceRRTMotionPlanner) getExtensionCandidate(
 	ctx context.Context,
 	randPosNode node,
 	ptgNum int,
 	curPtg tpspace.PTG,
-	rrt *rrtParallelPlannerShared,
+	rrt rrtMap,
 	nearest node,
-	planOpts *plannerOptions, // Need to pass this in explicitly for smoothing
-) *candidate {
-	nm := &neighborManager{nCPU: mp.planOpts.NumThreads}
+	invert bool,
+) (*candidate, error) {
+	nm := &neighborManager{nCPU: mp.planOpts.NumThreads / len(mp.tpFrame.PTGs())}
 	nm.parallelNeighbors = 10
 
 	var successNode node
 	// Get the distance function that will find the nearest RRT map node in TP-space of *this* PTG
 	ptgDistOpt := mp.algOpts.distOptions[curPtg]
+	if invert {
+		ptgDistOpt = mp.algOpts.invertDistOptions[curPtg]
```

Comment:
I understand why the invert bool exists but it seems really hacky and I wonder if theres a better way to do this.  Can't we just get the invert distance by sampleNode->treeNode and get normal distance by going treeNode->sample?  So instead of adding a bool to these functions just change the order of the arguments being passed in?

Reply:
I agree this is really hacky. I'll brainstorm different ways to do this.

> Can't we just get the invert distance by sampleNode->treeNode and get normal distance by going treeNode->sample?

That's what I originally implemented; unfortunately it doesn't work, as that's not quite how the problem needs to be formulated, due to the directionality of PTGs. You can't just invert the pose; you have to say "I need a pose X, such that when I compose pose A with the inverse of pose X, yields something close to pose B".

There may be a way to get it to fit together differently but my first three attempts failed.